File: src/main/java/helper/WeightedList.java
package helper;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Stream;

public class WeightedList<T extends Serializable> implements Serializable {

	private final ArrayList<Pair<Double, T>> values;

	public WeightedList() {
		// empty weights list
		this.values = new ArrayList<>();
	}

	/** Creates an evenly weighted group of size count. */
	public WeightedList(List<T> values) {
		this.values = new ArrayList<>();
		for (T value : values) {
			this.values.add(new Pair<>(1.0 / values.size(), value));
		}
	}

	/** Normalizes this range, setting the sum of the weights to 1.0.
	 * Note this returns a new weights list and doesn't modify the old one. */
	public WeightedList<T> normalize() {
		// divide each by the sum
		double sum = values.stream().mapToDouble(i -> i.one).sum();
		WeightedList<T> returned = new WeightedList<>();
		for (Pair<Double, T> i : this.values) {
			returned.values.add(new Pair<>(i.one / sum, i.two));
		}
		return returned;
	}

	/** returns the weight for the element at index. */
	public double getWeight(int index) {
		return values.get(index).one;
	}

	/** Sets the weight at index. Note this requires normalization afterwards. */
	public void setWeight(int index, double value) {
		values.set(index, new Pair<>(value, values.get(index).two));
	}

	/** Adds a weight to this list. */
	public void addWeight(double weight, T value) {
		values.add(new Pair<>(weight, value));
	}

	/** Returns the number of elements in this weighted list. */
	public int size() {
		return values.size();
	}

	/** Creates a stream of these elements. */
	public Stream<Pair<Double, T>> stream() {
		return values.stream();
	}

	/** Creates and returns a weighted list of the elements where the condition passes. */
	public WeightedList<T> valuesWith(Predicate<T> condition) {
		WeightedList<T> val = new WeightedList<>();
		for (Pair<Double, T> i : values) {
			if (condition.test(i.two)) {
				val.addWeight(i.one, i.two);
			}
		}
		return val;
	}

	/** Returns the total weight of this list. */
	public double totalWeight() {
		return values.stream().mapToDouble(i -> i.one).sum();
	}

	/** Returns the pair at the index. */
	public Pair<Double, T> get(int index) {
		return values.get(index);
	}

	/** Returns the item at the index. */
    public T getItem(int index) {
		return values.get(index).two;
    }
}

File: src/main/kotlin/com/zachjones/languageclassifier/entities/WeightedList.kt
package com.zachjones.languageclassifier.entities

import java.io.Serializable
import java.util.function.Predicate
import java.util.stream.Stream

class WeightedList<T> : Serializable {
    private val values: ArrayList<Pair<Double, T>>

    constructor() {
        // empty weights list
        this.values = ArrayList()
    }

    /** Creates an evenly weighted group of size count.  */
    constructor(values: List<T>) {
        this.values = ArrayList()
        for (value in values) {
            this.values.add(Pair(1.0 / values.size, value))
        }
    }

    /** Normalizes this range, setting the sum of the weights to 1.0.
     * Note this returns a new weights list and doesn't modify the old one.  */
    fun normalize(): WeightedList<T> {
        // divide each by the sum
        val sum = values.stream().mapToDouble { i: Pair<Double?, T> -> i.first!! }.sum()
        val returned = WeightedList<T>()
        for (i in this.values) {
            returned.values.add(Pair(i.first/ sum, i.second))
        }
        return returned
    }

    /** returns the weight for the element at index.  */
    fun getWeight(index: Int): Double {
        return values[index].first
    }

    /** Sets the weight at index. Note this requires normalization afterwards.  */
    fun setWeight(index: Int, value: Double) {
        values[index] = Pair<Double, T>(value, values[index].second)
    }

    /** Adds a weight to this list.  */
    fun addWeight(weight: Double, value: T) {
        values.add(Pair(weight, value))
    }

    /** Returns the number of elements in this weighted list.  */
    fun size(): Int {
        return values.size
    }

    /** Creates a stream of these elements.  */
    fun stream(): Stream<Pair<Double, T>> {
        return values.stream()
    }

    /** Creates and returns a weighted list of the elements where the condition passes.  */
    fun valuesWith(condition: Predicate<T>): WeightedList<T> {
        val values = WeightedList<T>()
        for (i in this.values) {
            if (condition.test(i.second)) {
                values.addWeight(i.first, i.second)
            }
        }
        return values
    }

    /** Returns the total weight of this list.  */
    fun totalWeight(): Double {
        return values.stream().mapToDouble { i: Pair<Double, T> -> i.first }.sum()
    }

    /** Returns the pair at the index.  */
    operator fun get(index: Int): Pair<Double, T> {
        return values[index]
    }

    /** Returns the item at the index.  */
    fun getItem(index: Int): T {
        return values[index].second
    }
}
